sdk/storage/Azure.Storage.Files.DataLake/src/DataLakeFileSystemClient.cs

```diff
@@ -2378,5 +2378,241 @@ private void SetNameFieldsIfNull()
+                    nameof(DataLakeFileSystemClient),
+                    message:
+                    $"{nameof(Uri)}: {Uri}\n" +
+                    $"{nameof(continuation)}: {continuation}\n" +
+                    $"{nameof(maxResults)}: {maxResults})");
+                try
+                {
+                    Response<FileSystemListBlobsHierarchySegmentResult> response = await DataLakeRestClient.FileSystem.ListBlobsHierarchySegmentAsync(
+                        clientDiagnostics: ClientDiagnostics,
+                        pipeline: Pipeline,
+                        resourceUri: _blobUri,
+                        delimiter: default,
+                        version: Version.ToVersionString(),
+                        prefix: path,
+                        marker: continuation,
+                        maxResults: maxResults,
+                        showonly: ListBlobsShowOnly.Deleted,
+                        async: async,
+                        operationName: operationName,
+                        cancellationToken: cancellationToken)
+                        .ConfigureAwait(false);
+
+                    return Response.FromValue(
+                        new PathDeletedSegment
+                        {
+                            Continuation = response.Value.Body.NextMarker,
+                            DeletedPaths = response.Value.Body.BlobItems.Select(blobItem => blobItem.ToPathHierarchyDeletedItem())
+                        },
+                        response.GetRawResponse());
+                }
+                catch (Exception ex)
+                {
+                    Pipeline.LogException(ex);
+                    throw;
+                }
+                finally
+                {
+                    Pipeline.LogMethodExit(nameof(DataLakeFileSystemClient));
+                }
+            }
+        }
+        #endregion Get Deleted Paths
+
+        #region Restore Path
+        /// <summary>
+        /// Restores a soft deleted path.
+        /// </summary>
+        /// <param name="deletedPath">
+        /// Required.  The path of the deleted path.
+        /// </param>
+        /// <param name="deletionId">
+        /// Required.  The deletion ID associated with the soft deleted path.
+        /// You can get soft deleted paths and their assocaited deletion IDs with <see cref="GetDeletedPathsAsync"/>.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{DataLakePathClient}"/> pointed at the newly
+        /// restored path.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Response<DataLakePathClient> RestorePath(
```

Comment:
I talked to Kamil, and we chose to go with `DataLakeFileSystemClient.RestorePath()` for now, instead of `.RestoreFile()` and `.RestoreDirectory`.

Also note that this API is on the `FileSystemClient`.  This is because you will be able to rename paths in the future when you restore them.  I followed the precedent of `BlobServiceClient.RestoreContainer()` and `ShareServiceClient.RestoreShare()`.